### ISSUE TITLE
Release 2023-02 follow-up

### ISSUE THREAD

**Issue by stonebig** [winpython/winpython#1215]
target: june 2023

wanted:
- pandas-2 (is pyarrow based)
- SQLalchemy-2
- python-3.11.4
- python-3.12.0 betas, with multi-interpreters early "experimental" APIs (package 'interpreters-3-12')

focus:
- barcoding eco-system (for GS1 barcodes): python-barcode, QRcode, opencv-python

exploratory:
- pycon US 2023 langage summit outcome:
  - [x] HPy: progress as a proposal, but not yet momentum
  - [x] nogil:  killed per cpython speed-up team
  - [x] per interpreter GIL  for Python-3.12.0b2: worked-out examples  at https://github.com/winpython/winpython_afterdoc/tree/master/docs/subinterpreters
- LargeLangageModel teaching per Andrew Ng:
   - 1 hour : [ChatGPT Prompt Engineering for Developers](https://learn.deeplearning.ai/chatgpt-prompt-eng)
   - 1 hour: [LangChain for LLM Application Development](https://learn.deeplearning.ai/langchain)
- LargeLangageModel on various WinPython problems: sans-Qt interface, build, clean-up, semi-automated DataScience ...
- a local LLM that could work on a gpu-less PC:
   - https://simonwillison.net/2023/May/5/mpt-7b/
   - https://github.com/ggerganov/llama.cpp
   - https://github.com/rustformers/llm
- Pyscript and JupyterLite progress.
- mamba ?

probably postponed:
- [ ] Jupyterlab-4, depends of extensions picking up very slowly ... summer slushy zone  ? https://github.com/jupyterlab/jupyterlab/issues/14590
- build winpython wheel out of make.py
- practice & build LLMs toolbox from factual learning examples on WinPython

assumptions:
- cPython multi-interpreters will greatly simplify multi-cpu usage,
- cPython will gain 10% speed-up per release,
- jupyterlite / pyodide is truly awesome.
- Rust is good for Python eco-system 
- swift inventor brainstorms mojo (a python+), because "Python [the syntax and the eco-sytem]" has won the world,
- only ".org" langages will remain: Python, Rust, WASM, Julia, C/C++, Javascript

**Comment by stonebig** [comment]
audit

**Comment by RoyiAvital** [comment]
Could we have MKL backed packages back?

**Comment by stonebig** [comment]
> Could we have MKL backed packages back?

there is no sign of cgohlke actively maintaining his wheels. So no, for now.

**Comment by Patrick-Cole** [comment]
Hi,

I notice that you did not update GDAL. Note that cgohlke is posting updates to some packages (including numpy+mkl) at: https://www.cgohlke.com/

What do you think?

**Comment by stonebig** [comment]
> Hi,
> 
> I notice that you did not update GDAL. Note that cgohlke is posting updates to some packages (including numpy+mkl) at: https://www.cgohlke.com/
> 
> What do you think?

Well, it's the upgrade path.

Is there any initiative in the domain or in GDAL to move, step-by-step to a more universal compile-friendly approach ? 
Suppose we remove GDAL

Current sufferings are:
- big official wheels not provided per their big team of maintainers, like GDAL
- and on the opposite, no big official wheels, like STAN
 
Solutions are:
- other package with the features and not the problem
- the package moving step-by-step to a more universal compile-friendly approach
- some eco-system progress helping
- some big names of the GAFAM tackling the issue

**Comment by stonebig** [comment]
remember to read what it implies

![image](https://github.com/winpython/winpython/assets/4312421/0f8f1256-cb07-4664-9a0f-1c8624ea106f)

**Comment by Patrick-Cole** [comment]
It's certainly food for thought. I certainly would not like to see GDAL removed and totally agree that it is not friendly in terms of providing convenient wheels (at least not for me). I currently don't see an alternative in this case. It's used by so many packages in addition to python that it's become an industry standard. 

I just thought that based on your answer on MKL backed packages, you might not be aware that cgohlke has resumed maintaining some of his wheels. For example:

https://github.com/cgohlke/geospatial-wheels/ has newer versions of GDAL and related libraries.

and

https://github.com/cgohlke/numpy-mkl-wheels/

Of course, if anything happens again to Christoph Gohlke, we will be back to square one, highlighting the weakness in this solution.

However, we certainly live in exciting times. It will be interesting to see where technology goes in the future and how python evolves with it.

**Comment by stonebig** [comment]
I am aware. The bus factor did apply about 9 month long last year, and  I see that come-back as a grace period.
And so the reluctance to re-use numpy-mkl, as we survived that.
And for the alternativeless GDAL, the clock is ticking for a survival plan.

**Comment by stonebig** [comment]
On python speed-up and the new implementation "Mojo": https://www.youtube.com/watch?v=pdJQ8iVTwj8&t=11939s
to summarize:
- Python (the langage, the community) won (said by the inventor of LLVM and Swift)
- Mojo intends to be a super-set of the langage that can be up to 8 000x faster:
    - 20x: compiled
    - 20x: runs in parallel
    - 20x: can run on gpus
    - includes an cpython for non-trivial situations or not tweaked python code

**Comment by stonebig** [comment]
prelude: python-3.12.0b2 (2023-06-1x)

This is the basic winpythondot, but with three examples of multi-interpreters in notebooks/docs/subinterpreters.
As very fresh matter, it doesn't work syntacicaly as stated in Eric Snow demos...

simplest example:
````
import _xxsubinterpreters as interpreters
from threading import Thread
import textwrap as tw

def test_threading_interpreter():
    interp = interpreters.create()

    def t():
        interpreters.run_string(interp,tw.dedent("""
            import time
            for _ in range(50):
                print(end='.', flush=True)
                time.sleep(0.05)
            print("End of interp job")
        """))

    thread = Thread(target = t)
    print(f"Before thread.start(), {interpreters.is_running(interp) = }")
    thread.start()
    print(f"After thread.start(), {interpreters.is_running(interp) = }")
    thread.join()
    print(f"After thread.join(), {interpreters.is_running(interp) = }")

if __name__ == "__main__":
    test_threading_interpreter()
````

**Comment by stonebig** [comment]
prelude: python-3.12.0b2b (2023-06-1x)

Change from python-3.12.0b2:
- Infrastructure
- New Packages:
  - interpreters-3-12 (to use as "import interpreters" on  Eric Snow presentations)

**Comment by hiccup7** [comment]
A comment in this issue provides a work-around for a bug in Spyder v5.4.3:
https://github.com/spyder-ide/spyder/issues/20871
The bug doesn't exist in WinPython 3.10.9.0 that contains Spyder v5.4.2.

**Comment by stonebig** [comment]
> A comment in this issue provides a work-around for a bug in Spyder v5.4.3: [spyder-ide/spyder#20871](https://github.com/spyder-ide/spyder/issues/20871) The bug doesn't exist in WinPython 3.10.9.0 that contains Spyder v5.4.2.

it was corrected starting  ipython-8.12.2, and we shipped ipython-8.13.1.

**Comment by hiccup7** [comment]
mpmath v1.3.0 was released 2023 March 7.  https://pypi.org/project/mpmath/
It was tested on Python 3.11.  mpmath is a dependency of sympy.
Another release is planned in a few days with a large number of changes, including support for Python 3.12.

**Comment by hiccup7** [comment]
Numba v0.57.1 was released 2023 June 21.  https://github.com/numba/numba/blob/main/CHANGE_LOG
It fixes regressions with v0.57.0.

**Comment by RoyiAvital** [comment]
Hi, It should have been release on June.
Could we assume to have it any day now?

**Comment by stonebig** [comment]
will try

**Comment by stonebig** [comment]
Meta seems determined to make Nogil happen
![image](https://github.com/winpython/winpython/assets/4312421/dc7c48f6-bb66-4011-86cd-66530b9160bb)

**Comment by stonebig** [comment]
rc (2023-07-09)

Changes from WinPython 2023-02 b1:
- Infrastructure:
   - Python-3.12.0b3
   - build: move to packages.ini to utf-8
- New Packages:
- Upgraded Packages: 
   - jupyterlab-3.6.5, matplotlib-3.7.2
   - datashader-0.15.1, numba-0.57.1
   - duckdb-0.8.1, polars-0.18.4
- variation per version:
   - interpreters-3-12 is only for python-3.12.0b3

Areas of particular interest for testers:
- all stable


 MD5                             | SHA-1                                    | SHA-256                                                          | Binary                            | Size               | blake2b-256                                                     
---------------------------------|------------------------------------------|------------------------------------------------------------------|-----------------------------------|--------------------|-----------------------------------------------------------------

**Comment by stonebig** [comment]
final (2023-07-15)

Hello,

As no major complains occured, rc binaries are now re-tagged as final without any change.

 MD5                             | SHA-1                                    | SHA-256                                                          | Binary                            | Size               | blake2b-256                                                    
---------------------------------|------------------------------------------|------------------------------------------------------------------|-----------------------------------|--------------------|-----------------------------------------------------------------
83b5c99aa9303f93c06241928a1ef51c | 8c28ab41e68f53faa081530324c26dbe7b028935 | 09dacbf391adb11ace34eb90cd02ae3fc3749df612b8a1fd288e7e5c47bc1973 | Winpython64-3.11.4.0dot.exe       |   31 541 585 Bytes | ea780b4c7781dda507a09fa7ef856fe58b2f49959f849f3f1a01106d5fda35fe
6d9f7fed5327257aa86be3b7caa6916d | 11dbfa8c2ec1d225bdb0c6f19f75d62c12e640c4 | 193233967bd2b6a6cd705ac3b17dac9c4864ae8e56744a2b89caed91f2fb6f05 | Winpython64-3.11.4.0.exe          |  785 746 334 Bytes | feba52aca59c85f8efce1ee58fd28c052f7505966728268481f9405956e70f41

re-tag binaries are to be found at  :
- https://github.com/winpython/winpython/releases/tag/6.4.20230625final
- https://sourceforge.net/projects/winpython/files/WinPython_3.11/3.11.4.0/

**Comment by stonebig** [comment]
Because it's possible. testing Duckdb on the SQL sudoku exercize:
- DuckDB takes 17 seconds vs SQLite taking 9 seconds,
- it's not bad given SQL is not designed to solve sudokus, and duckDB is not targeted to do algorithms in simple SQL
- still it would have been extraordinar if DuckDB had been quicker  ...
- ... maybe it's till possible by tweaking the SQL to DuckDB features ?
```
import duckdb

con = duckdb.connect()
# question is  '417369825632.........7......2.....6.....8.4......1.......6.3.7.5..2.....1.4......'
# answer is    '417369825632158947958724316825437169791586432346912758289643571573291684164875293'


sql="""WITH RECURSIVE input(sud) AS (VALUES(
  '417369825632.........7......2.....6.....8.4......1.......6.3.7.5..2.....1.4......'
))
, digits(z, lp) AS (VALUES('1', 1) UNION ALL SELECT CAST(lp+1 AS TEXT), lp+1 FROM digits WHERE lp<9)
, x(s, ind) AS (SELECT sud, instr(sud, '.') FROM input
 UNION ALL
 SELECT substr(s, 1, ind-1) || z || substr(s, ind+1), instr( substr(s, 1, ind-1) || z || substr(s, ind+1), '.' ) FROM x
 , digits AS z WHERE ind>0 AND NOT EXISTS (SELECT 1 FROM digits AS lp
  WHERE z.z = substr(s, ((ind-1)//9)*9 + lp, 1) OR z.z = substr(s, ((ind-1)%9) + (lp-1)*9 + 1, 1)
  OR z.z = substr(s, (((ind-1)//3) % 3) * 3 + ((ind-1)//27) * 27 + lp + ((lp-1) // 3) * 6, 1))) SELECT s FROM x WHERE ind=0;
  """

results = duckdb.sql(sql).fetchall()

print(results)

**Comment by stonebig** [comment]
tweaking syntax has no effect on speed


sql="""WITH RECURSIVE input(sud) AS (VALUES(
  '417369825632.........7......2.....6.....8.4......1.......6.3.7.5..2.....1.4......'
))
, digits(z, lp) AS (VALUES('1', 1) UNION ALL SELECT CAST(lp+1 AS TEXT), lp+1 FROM digits WHERE lp<9)
, x(s, ind) AS (SELECT sud, instr(sud, '.') FROM input
 UNION ALL
 SELECT substr(s, 1, ind-1) || z || substr(s, ind+1)
 , instr( substr(s, 1, ind-1) || z || substr(s, ind+1), '.' )
 FROM x
 , digits AS z WHERE ind>0 AND NOT EXISTS (SELECT 1 FROM digits AS lp
  WHERE
  z.z = s[((ind-1)//9)*9 + lp]
  OR
  z.z = s[((ind-1)%9) + (lp-1)*9 + 1]
  OR z.z = s[(((ind-1)//3) % 3) * 3 + ((ind-1)//27) * 27 + lp + ((lp-1) // 3) * 6])) SELECT s FROM x WHERE ind=0;
  """
